src/mumble/JackAudio.cpp

```diff
@@ -0,0 +1,618 @@
+
+	if (client) {
+		in_port = jack_port_register(client, "input", JACK_DEFAULT_AUDIO_TYPE, JackPortIsInput, 0);
+		if (in_port == NULL) {
+			qWarning("JackAudioSystem: unable to register 'input' port");
+			close_jack();
+			return;
+		}
+
+		bJackIsGood = true;
+		iBufferSize = jack_get_buffer_size(client);
+		iSampleRate = jack_get_sample_rate(client);
+
+		setNumberOfOutPorts(iOutPorts);
+
+		if (bJackIsGood == false) {
+			close_jack();
+			return;
+		}
+
+		err = jack_set_process_callback(client, process_callback, this);
+		if (err != 0) {
+			qWarning("JackAudioSystem: unable to set process callback - jack_set_process_callback() returned %i", err);
+			close_jack();
+			return;
+		}
+
+		err = jack_set_sample_rate_callback(client, srate_callback, this);
+		if (err != 0) {
+			qWarning("JackAudioSystem: unable to set sample rate callback - jack_set_sample_rate_callback() returned %i", err);
+			close_jack();
+			return;
+		}
+
+		err = jack_set_buffer_size_callback(client, buffer_size_callback, this);
+		if (err != 0) {
+			qWarning("JackAudioSystem: unable to set buffer size callback - jack_set_buffer_size_callback() returned %i", err);
+			close_jack();
+			return;
+		}
+
+		jack_on_shutdown(client, shutdown_callback, this);
+
+		// If we made it this far, then everything is okay
+		qhInput.insert(QString(), tr("Hardware Ports"));
+		qhOutput.insert(QString::number(1), tr("Mono"));
+		qhOutput.insert(QString::number(2), tr("Stereo"));
+		bJackIsGood = true;
+
+	} else {
+		QStringList errors = jackStatusToStringList(status);
+		qWarning("JackAudioSystem: unable to open jack client due to %i errors:", errors.count());
+		for (int i = 0; i < errors.count(); ++i) {
+			qWarning("JackAudioSystem:  %s", qPrintable(errors.at(i)));
+		}
+		bJackIsGood = false;
+		client = NULL;
+	}
+}
+
+void JackAudioSystem::close_jack() {
+
+	QMutexLocker lock(&qmWait);
+	if (client) {
+		int err = 0;
+		err = jack_deactivate(client);
+		if (err != 0)  {
+			qWarning("JackAudioSystem: unable to remove client from the process graph - jack_deactivate() returned %i", err);
+		}
+
+		bActive = false;
+
+		if (in_port != NULL) {
+			err = jack_port_unregister(client, in_port);
+			if (err != 0)  {
+				qWarning("JackAudioSystem: unable to unregister in port - jack_port_unregister() returned %i", err);
+			}
+		}
+
+		for (unsigned i = 0; i < iOutPorts; ++i) {
+			if (out_ports[i] != NULL) {
+				err = jack_port_unregister(client, out_ports[i]);
+				if (err != 0)  {
+					qWarning("JackAudioSystem: unable to unregister out port - jack_port_unregister() returned %i", err);
+				}
+			}
+		}
+
+		err = jack_client_close(client);
+		if (err != 0) {
+			qWarning("JackAudioSystem: unable to to disconnect from the JACK server - jack_client_close() returned %i", err);
+		}
+
+		delete [] output_buffer;
+		output_buffer = NULL;
+
+		client = NULL;
+	}
+	bJackIsGood = false;
+}
+
+
+void JackAudioSystem::auto_connect_ports()
+{
```

Comment:
The bracket should be at the end of the line above (to match the style across the file).

Reply:
Done